### ISSUE TITLE
overwriting modules in TYPE_CHECKING is ignored when module is used in class

### ISSUE THREAD

**Issue by MrHenning** [microsoft/pyright#6300]
I am using the Python library [pyqtgraph](https://github.com/pyqtgraph/pyqtgraph) to build a GUI, with `pyqtgraph` using `PyQt5` as a GUI backend.
Python type-checkers don't recognize pyqtgraph's imports of Qt elements:
```python
from pyqtgraph.Qt import QtWidgets
out = QtWidgets.QMessageBox()
``` 
```
test.py:2:26 - error: "QMessageBox" is not a known member of module "QtWidgets" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

So, to be able to type-check Qt, I install [PyQt5-stubs](https://github.com/python-qt-tools/PyQt5-stubs) and use PyQt's modules for type-checking:
```python
from typing import TYPE_CHECKING
from pyqtgraph.Qt import QtWidgets
if TYPE_CHECKING:
    from PyQt5 import QtWidgets
    
out = QtWidgets.QMessageBox()
``` 
```
0 errors, 0 warnings, 0 informations 
```

So, all good.

However, if I put my code in a class, pyright complains:
```python
from typing import TYPE_CHECKING
from pyqtgraph.Qt import QtWidgets
if TYPE_CHECKING:
    from PyQt5 import QtWidgets

class Foo:
    def __init__(self) -> None:
        out = QtWidgets.QMessageBox()
```
```
test.py:8:25 - error: "QMessageBox" is not a known member of module "QtWidgets" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

Am I missing something or is this a bug?

Versions:
Python 3.11.6 
pyright v1.1.333
pyqtgraph                 0.13.3
pyqt                      5.15.9          
pyqt-stubs                5.15.6.0

**Comment by erictraut** [comment]
Yes, this behavior is intended. You are assigning two values to the symbol `QtWidgets` on two code paths in the module scope. If you reference this symbol in the module path (or directly within a class body), pyright can determine which one you mean based on code flow analysis. When you reference the symbol from an independent inner scope (such as the `__init__` method in your example), it can no longer determine which type `QtWidgets` will have at the time that code runs, so its type is inferred based on the union of all values assigned to it. In your case, you're assigning values to it in two different import statements. If you want to make it unambiguous for the type checker, you need to change your code to the following:

```python
if TYPE_CHECKING:
    from PyQt5 import QtWidgets
else:
    from pyqtgraph.Qt import QtWidgets
```

**Comment by MrHenning** [comment]
That makes sense!
Thanks a lot for the detailed explanation, and also how to fix it in my code 🙂
